### ISSUE TITLE
Spec compliance: require a request body on `POST /rooms/{roomId}/receipt/{receiptType}/{eventId}`

### ISSUE THREAD

**Issue by reivilibre** [matrix-org/synapse#10534]
We currently don't require a body on `POST /rooms/{roomId}/receipt/{receiptType}/{eventId}`.

Initially, this seems to have been a mistake, but since #10531 [v1.40.0rc2], it's a workaround because Element Android relied on that behaviour (which we broke by actually using the body for hidden read receipts in #10413 [v1.40.0rc1]).

https://github.com/vector-im/element-android/issues/3789 is the issue on Element Android.

Eventually, we should return to enforcing the spec.

**Comment by richvdh** [comment]
I'd like us to set a timescale for deprecation of this workaround while it's still fresh in our heads.

**Comment by callahad** [comment]
Question: Do we constrain the workaround per above, or do we wait even longer? The rate of decay is slow enough that I don't think we'll dip under 10% for many months.

**Comment by callahad** [comment]
> it feels nasty to make the homeserver have to discriminate on user agent

Agreed that a UA-specific hack is... a hack... but at least it stops new callers from accidentally adopting the deprecated behavior, so it at least ensures that the ecosystem does not regress.

> it's not particularly fair

I think it's fair _enough._ To the best of our knowledge (which includes seeing people yell when we broke this), _only_ Element Android and its forks have ever been subject to this particular misimplementation of the spec. From looking at _all_ clients that advertised themselves as "android" on October 13th, limiting the hack to _just_ old versions of Riot, Element Android, and SchildiChat would get us to a point where _literally_ more than 99.9% of all Android clients we saw on matrix.org would explicitly be known to work.

Basically, I'm willing to risk that long tail of 0.1% of android clients which we've not explicitly verified whether they comply with the spec or not.

**Comment by callahad** [comment]
Next step is to rip out the hack once we're comfortable with ecosystem adoption...

Dan to pull numbers again.

**Comment by callahad** [comment]
(Waiting until next week Wednesday just so we get a more representative sample of people coming back from holidays)

**Comment by clokep** [comment]
FWIW #12168 is an update to the hidden read receipt logic which means we don't actually need a body there anymore. We could remove all the code that cares about the body or still move forward with this?

**Comment by erikjohnston** [comment]
Android team are happy for us to drop the workaround patch now!

**Comment by DMRobertson** [comment]
> Android team are happy for us to drop the workaround patch now!

In that case, next step is to strip out the user-agent specific logic added in https://github.com/matrix-org/synapse/pull/11157, and write a removal notice in the changelog.
